### PR TITLE
feat: implement RangeManipulate

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Implement `RangeManipulate" plan, which collapses range vectors for range function to calculate. Section in RFC: https://github.com/GreptimeTeam/greptimedb/blob/develop/docs/rfcs/2022-12-20-promql-in-rust/rfc.md#plan-and-expression 

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
#596

### CODE REVIEW

File: src/promql/src/extension_plan/range_manipulate.rs [hunk 1]
```diff
@@ -0,0 +1,504 @@
...
+            let mut range_end = 0;
+            for (index, ts) in ts_column.values().iter().enumerate() {
+                if ts + self.range >= curr_ts {
+                    range_start = range_start.min(index);
+                }
+                if *ts <= curr_ts {
+                    range_end = range_end.max(index);
+                }
+            }
+            result.push((range_start as _, (range_end + 1 - range_start) as _));
```
Comment (evenyag) [inline src/promql/src/extension_plan/range_manipulate.rs:376]: What if no timestamp could be found in a range? Do we need to check `range_start <= range_end` here?

Comment (waynexia) [reply src/promql/src/extension_plan/range_manipulate.rs]: return an empty range `(0,0)`, nice catch

**Review by evenyag — APPROVED** [review]

File: src/promql/src/extension_plan/range_manipulate.rs [hunk 2]
```diff
@@ -0,0 +1,527 @@
...
+                Poll::Ready(batch.map(|batch| batch.and_then(|batch| self.manipulate(batch))))
+            }
+            Poll::Pending => Poll::Pending,
+        };
+        self.metric.record_poll(poll)
+    }
+}
+
+impl RangeManipulateStream {
+    pub fn manipulate(&self, input: RecordBatch) -> ArrowResult<RecordBatch> {
```
Comment (killme2008) [inline src/promql/src/extension_plan/range_manipulate.rs:330]: Can we refer to the original source code in Prometheus for this logic?  Looks complicated.

**Review by killme2008 — APPROVED** [review]
LGTM

File: src/promql/src/extension_plan/range_manipulate.rs (hunk 2 above)
Comment (waynexia) [reply src/promql/src/extension_plan/range_manipulate.rs:333]: 👍 nice suggestion

**Pull request merged**